### ISSUE TITLE
How can we get all the repos?

### ISSUE THREAD

**Issue by zeke** [zeke/repos-using-electron#2]
TL:DR: How can we collect a complete list of all public repositories that depend on `electron-prebuilt`?

The Electron team is interested in collecting data about open-source apps and tools being built with Electron. To that end, we want to find all public repositories with a `package.json` file containing the `electron-prebuilt` package name in `dependencies` or `devDependencies`. We are currently using the libraries.io API to collect this data. See [fetch.js](https://github.com/zeke/repos-using-electron/blob/fe9ddce701773bfe36f6eaca55a340f221a2782c/fetch.js

The libraries.io API is maintained by a single person in his spare time, so it has some pagination bugs and occasional service disruptions. I was looking into using GitHub’s search API as an alternative, but it doesn't seem to be designed for this use case:

> Think of it the way you think of performing a search on Google. It's designed to help you **find the one result you're looking for** (or maybe the few results you're looking for). Just like searching on Google, you sometimes want to see a few pages of search results so that you can find the item that best meets your needs. To satisfy that need, the GitHub Search API provides up to **1,000 results for each search.**

According to libraries.io, there are nearly 5000 repos depending on `electron-prebuilt`:

```
curl -i -s "https://libraries.io/api/npm/electron-prebuilt/dependent_repositories/?per_page=1" | grep Total

Total: 4946
```

The 1000-result limit aside, the GitHub API is telling me I need to scope my search to "at least one user, organization, or repository", which doesn't make sense for my query:

```
curl -n "https://api.github.com/search/code?q=electron-prebuilt+in:file+filename:package.json"  
{
  "message": "Validation Failed",
  "errors": [
    {
      "message": "Must include at least one user, organization, or repository",
      "resource": "Search",
      "field": "q",
      "code": "invalid"
    }
  ],
  "documentation_url": "https://developer.github.com/v3/search/#search-code"
}
````

Any ideas? Thanks for any and all input.

**Comment by zeke** [comment]
Got some answers from @chrismwendt:

> You can run that query from the web UI because that scoping limitation only applies to the API
https://github.com/search?utf8=%E2%9C%93&q=electron-prebuilt+in:file+filename:package.json

> We're working on enabling global code search through the API https://github.com/github/github/issues/43715

> I wish we supported this kind of "scan" functionality so you didn't have to go through the UI. All I would say is don't submit many requests per second, and don't request really high page numbers, but we have limits for both of those, so you're probably fine no matter what you do.

I asked:

> is there an undocumented query param for changing the number of results per page? And is `number_of_results/results_per_page` (in this case, `10,024/10`) considered a “really high number"?

> Also noticing a different number of results each time I refresh the page, or go beyond page 1. Is that expected behavior?

To which @chrismwendt responded:

> I don't know of any such param to change # of results per page

> We have a query timeout which returns whatever results that Elasticsearch found before the timer is up, so you're probably seeing nondeterminism there

> Another problem is that you might see files twice, or miss files, since these are separate requests and the data can change between requests

> Since this "scan" functionality is requested so often, @rafer (on the data engineering team) is working on implementing it. It's unclear right now whether or not it will become a public API

**Comment by andrew** [comment]
I'm working on Libraries full time now, do feel free to let me know how I can help with any research related tasks 🎉

**Comment by zeke** [comment]
@andrew! That would be awesome. I plan to focus on this project in earnest next week. I'll ping you. :)

**Comment by rafer** [comment]
@zeke had any luck getting vpn access yet? (so I can get the packages file)

**Comment by zeke** [comment]
<img width="974" alt="screen shot 2017-02-05 at 8 49 52 am" src="https://cloud.githubusercontent.com/assets/2289/23239359/8eab7f1c-f91b-11e6-8486-ffa46f41606c.png">
